pkg/apis/core/validation/validation.go

```diff
@@ -2092,6 +2092,30 @@ func validateDataSource(dataSource *core.TypedLocalObjectReference, fldPath *fie
 	return allErrs
 }
 
+// validateDataSource2 validates a DataSourceRef2 in a PersistentVolumeClaimSpec
+func validateDataSource2(dataSource2 *core.TypedCrossNamespaceObjectReference, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(dataSource2.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), "CrossNamespaceSourceProvisioning Name cannot be empty"))
+	}
+	if len(dataSource2.Kind) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("kind"), "CrossNamespaceSourceProvisioning kind cannot be empty"))
+	}
+	if dataSource2.Kind != "VolumeSnapshot" {
```

Comment:
I have some discomfort with the idea of embedding a namespace field directly into the PVC object. I agree that it's "cleaner" than the alternative, but given Kubernetes' history of locking pods/pvc/snapshots into a single namespace I wonder if it wouldn't be better to force cross-namespace use cases to have an extra layer of indirection to break out of this historical limitation.
The backwards-compatibility concerns notwithstanding, it seems like a fairly small amount of extra work to create a new object that bridges the local PVC to the remote snapshot, and it seems appropriate given that we probably want namespaces to continue to be a meaningful boundary in user's minds.
When I add in the concerns about backwards compatibility, I worry it might not even be possible to straightforwardly expand the existing field and we might be forced to add yet another field (need to think this through) and if it comes to that then just having an extra level of indirection seems like the cleaner option.

Reply:
This is kind of what I had in mind:
```
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: foreign-snapshot
  namespace: bar
spec:
  source:
    volumeSnapshotContentName: baz
---
apiVersion: snapshotref.storage.k8s.io/v1alpha1
kind: SnapshotRef
metadata:
  name: ref
  namespace: foo
foreignSnapshot:
  name: foreign-snapshot
  namespace: bar
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: local-pvc
  namespace: foo
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Mi
  dataSourceRef:
    apiGroup: snapshotref.storage.k8s.io/v1alpha1
    kind: SnapshotRef
    name: ref
```

Reply:
I think this is similar to the `VolumeSnapshotLink` in the earlier version of the KEP.  I also like this approach better.

Reply:
There will be 4 approaches below:

(a) Add a new `VolumeSnapshotLink` CRD  (earlier version of the KEP)
  - Pros:
    - No backward compatibility issue happens for existing data source behaviors
  - Cons:
    - Requires users to learn an additional CRD

(b) Add a new `DataSourceRef2` field to `PersistentVolumeClaim` (previous implementation (see [branch](https://github.com/ttakahashi21/kubernetes/tree/BK202210031-KEP-3294)))
  - Pros:
    - Doesn't require users to learn an additional CRD
  - Cons:
    - Requires users to learn an additional dataSource type
    - Needs to consider consistency on handling multiple dataSource types

(c) Add a new `namespace` field to `DataSourceRef` field in `PersistentVolumeClaim` (current implementation)
  - Pros:
    - Doesn't require users to learn an additional CRD nor an additional dataSource type
  - Cons:
    - `DataSource` and `DataSourceRef` will mismatch when namespace is specified in `DataSourceRef`, which is different from existing `DataSourceRef` beta behavior

(d) Add a new `namespace` field to both `DataSourceRef` field and `DataSource` field in `PersistentVolumeClaim`
  - Pros:
    - The same as (c)
    - Will be able to just copy the fields between `DataSource` and `DataSourceRef`
  - Cons:
    - Requires changing the DataSource field which is already GA
  
 Because (d) will require an obvious breaking change, I've implemented (c) to compare with (a) and (b). 
Let's check if changes in (c) will be acceptable and decide which approach to take.

Note that the logic of (c) will be like:

- `CrossNamespaceVolumeDataSource` feature is disabled:
  - Drop the contents of the dataSourceRef field if `dataSourceRef.Namespace` is specified
  - Always copy between `DataSource` and `DataSourceRef` if `AnyVolumeDataSource` feature gate is enabled

- `CrossNamespaceVolumeDataSource` feature is enabled:
  - Always copy from `DataSource` to `DataSourceRef` if `AnyVolumeDataSource` feature gate is enabled 
  - Copy from `DataSourceRef` to `DataSource` if `AnyVolumeDataSource` feature gate is enabled and `DataSourceRef.Namespace` == ""
  - Reject `DataSource` if `DataSourceRef.Namespace` != ""
  - Reject `DataSourceRef` if `DataSourceRef.Namespace` != "" and specified resource isn't snapshot (until we support generic resource in this feature)
---

Reply:
I think that in order to use the new features, users should be using `dataSourceRef` and not `dataSource`.I can understanding copying `dataSource` to `dataSourceRef` but not the other way around. @bswartz do you remember why we copy `dataSourceRef` to `dataSource`? Doesn't that break the semantics of the old `dataSource` field? (the reason why we created a new `dataSourceRef` field)?

Reply:
Discussed with Ben offline. The reason why we copy `dataSourceRef` to `dataSource` is because we want the older provisioners to return an error when an unsupported datasource is specified

I think for adding namespace support, it is reasonable that controllers need to be updated to understand the new behavior. So the proposed solution of not copying DataSourceRef if namespace is set sounds good to me.